### PR TITLE
Allow proper multi-controller support in H2V (Re)

### PR DESCRIPTION
**Pull request by Himanshu-01**
# Includes
* Adds `input_globals` and `input_abstraction_globals`
* Fixes some of the hired-gun patches that tied all inputs to `input_globals.gamepad_state[0]`
* User interfaces like pregame lobby can now detect multiple controller inputs successfully (distinctly)
* Multiple player inputs work ingame

# Todo List
- [x]  Completely breaks Imgui AdvancedSetting ControllerInput.cpp hacky patches that can crash
- [x]  `input_abstraction_update()` needs to be redone 
- [ ]  Certain UIs like main_menu dont like multiple local users
- [ ]  Connecting  2 or more controller breaks button hold (to fix in future)
- [x] Raw mouse Hack needs to be redone
- [x] Radial/Axial and Modern controller aiming needs to be redone

# Credits 
https://github.com/twist84/ManagedDonkey/tree/main/game/source/input

### CODE REVIEW

File: xlive/Blam/Engine/input/input_abstraction.h [hunk 1]
```diff
@@ -1,3 +1,217 @@
 #pragma once
 
+#include "controllers.h"
+#include "input_windows.h"
+
+#define K_MAXIMUM_NUMBER_OF_GAME_FUNCTION_BINDS 8
```
Comment (Berthalamew) [inline xlive/Blam/Engine/input/input_abstraction.h]: rename so this is all lowercase

File: xlive/Blam/Engine/input/input_abstraction.h [hunk 2]
```diff
@@ -1,3 +1,217 @@
...
+	_button_mouse_yaw_right = 0x32,
+	_button_mouse_pitch_forward = 0x33,
+	_button_mouse_pitch_backward = 0x34,
+	_extended_button_gamepad_yaw_left = 0x35,
+	_extended_button_gamepad_yaw_right = 0x36,
+	_extended_button_gamepad_pitch_forward = 0x37,
+	_extended_button_gamepad_pitch_backward = 0x38,
+
+	NUMBER_OF_BUTTON_KEYS = 0x39,
+	NUMBER_OF_ABSTRACT_BUTTONS = 0x2D,
```
Comment (Berthalamew) [inline xlive/Blam/Engine/input/input_abstraction.h:77]: rename these to use the k_ convention

File: xlive/Blam/Engine/input/input_abstraction.cpp [hunk 3]
```diff
@@ -1,7 +1,442 @@
...
+	real32 delta = fabs(angle - flt_7B9F78[index]);
+	real_angle min_delta = right_stick ? 0.17453292f : 0.61086524f;
+
+	if (delta >= min_delta)
+	{
+
+		float sign = 0.0f;
+		if (fabs(stick->yaw) <= fabs(stick->pitch))
+		{
+			if (stick->pitch >= 0.0f)
```
Comment (Berthalamew) [inline xlive/Blam/Engine/input/input_abstraction.cpp]: can use tertiary operators for these statements as well as the ones below

File: xlive/Blam/Engine/input/input_windows.h [hunk 4]
```diff
@@ -1,29 +1,163 @@
...
+	_xinput_gamepad_left_shoulder = 0x8,
+	_xinput_gamepad_right_shoulder = 0x9,
+	_xinput_gamepad_a = 0xa,
+	_xinput_gamepad_b = 0xb,
+	_xinput_gamepad_x = 0xc,
+	_xinput_gamepad_y = 0xd,
+	_xinput_gamepad_left_trigger = 0xe,
+	_xinput_gamepad_right_trigger = 0xf,
+
+	K_NUMBER_OF_XINPUT_BUTTONS = 0xE,
```
Comment (Berthalamew) [inline xlive/Blam/Engine/input/input_windows.h]: lowercase constant

File: xlive/Blam/Engine/input/input_windows.h [hunk 5]
```diff
@@ -1,29 +1,163 @@
 #pragma once
 #include "controllers.h"
 
+#define K_NUMBER_OF_WINDOWS_INPUT_VIRTUAL_CODES 256
```
Comment (Berthalamew) [inline xlive/Blam/Engine/input/input_windows.h]: lowercase constant

File: xlive/Blam/Engine/input/input_windows.h [hunk 6]
```diff
@@ -1,29 +1,163 @@
...
+	virtual void XUpdateState(void) override;
+	virtual uint32 XGetState(XINPUT_STATE* state) override;
+	virtual void XSetState(XINPUT_VIBRATION* state) { return; };
+	virtual void XUpdateImmediate(void) { return; };
+};
+ASSERT_STRUCT_SIZE(dinput_device, 0x40);
+
+
+
+#pragma pack(push,1)
```
Comment (Berthalamew) [inline xlive/Blam/Engine/input/input_windows.h]: remove this if it isn't needed

**Review by Berthalamew — CHANGES_REQUESTED** [review]

File: xlive/Blam/Engine/input/input_abstraction.cpp [hunk 7]
```diff
@@ -1,7 +1,442 @@
...
+void input_abstraction_update_default_throttle(point2d* thumb, real_euler_angles2d* stick)
+{
+	real32 pointy = (real32)thumb->y;
+	real32 pointx = (real32)thumb->x;
+	real_angle angle = atan2(pointy, pointx);
+
+	real32 magnitude = MAX(fabs(sin(angle)), fabs(cos(angle)));
+	real32 inverse_magnitude = 1.0 / magnitude;
+
+	real32 normalized_yaw = (real32)(pointx * inverse_magnitude) * 0.000030518509f;
```
Comment (nukeulater) [inline xlive/Blam/Engine/input/input_abstraction.cpp]: mulitply by 0.000030518509f or divide by 32.767f, division by 32.767 is clearer

File: xlive/Blam/Engine/input/input_abstraction.cpp [hunk 8]
```diff
@@ -1,7 +1,442 @@
...
+{
+	input_abstraction_update_default_throttle(&gamepad_state->thumb_left, left_stick);
+	input_abstraction_update_default_throttle(&gamepad_state->thumb_right, right_stick);
+
+	input_abstraction_post_update_all_throttles(left_stick, right_stick, &gamepad_state->thumb_left, &gamepad_state->thumb_right);
+}
+
+void input_abstraction_update_throttles_modern(s_gamepad_input_button_state* gamepad_state, real_euler_angles2d* left_stick, real_euler_angles2d* right_stick)
+{
+	left_stick->yaw = gamepad_state->thumb_left.x * 0.000030518509;
```
Comment (nukeulater) [inline xlive/Blam/Engine/input/input_abstraction.cpp]: [same here](https://github.com/pnill/cartographer/pull/631/files#r1573316666)

File: xlive/Blam/Engine/input/input_abstraction.h (hunk 2 above)
Comment (Himanshu-01) [reply xlive/Blam/Engine/input/input_abstraction.h:77]: its actually called NUMBER_OF_EXTENDED_CONTROL_BUTTONS

File: xlive/Blam/Engine/input/input_abstraction.cpp (hunk 3 above)
Comment (Himanshu-01) [reply xlive/Blam/Engine/input/input_abstraction.cpp]: used ternary for calculating sign

File: xlive/Blam/Engine/input/input_abstraction.cpp (hunk 8 above)
Comment (Himanshu-01) [reply xlive/Blam/Engine/input/input_abstraction.cpp]: Replaced constants with more meaningful constexpr

**Review by nukeulater — APPROVED** [review]

File: xlive/Blam/Engine/input/input_xinput.cpp [hunk 9]
```diff
@@ -1,66 +1,217 @@
...
+	}
+}
+
+int16 input_xinput_adjust_thumb_axis_deadzone(int16 thumb_axis, int16 thumb_deadzone)
+{
+	if (thumb_axis > thumb_deadzone)
+		return  0x7FFF * (thumb_axis - thumb_deadzone) / (0x7FFF - thumb_deadzone);
+
+	if (thumb_axis < -thumb_deadzone)
+		return ~0x7FFF * (thumb_axis + thumb_deadzone) / (~0x7FFF + thumb_deadzone);
```
Comment (nukeulater) [inline xlive/Blam/Engine/input/input_xinput.cpp:113]: ~0x7FFF (NOT 0x7FFF) is not -32767, but -32768. Should be -0x7FFF instead, unless it is intentional.

File: xlive/H2MOD/GUI/imgui_integration/MOTD.cpp [hunk 10]
```diff
@@ -165,7 +164,7 @@ namespace ImGuiHandler
 				ImTextureID texId = ImGuiHandler::GetTexture(patch_notes);
 				ImGui::Image(texId, ImGui::GetIO().DisplaySize);
 
-				if (ControllerInput::get_controller_input(0)[16] == 1
+				if (input_get_gamepad_state(0)->button_frames_down[_xinput_gamepad_a]
```
Comment (nukeulater) [inline xlive/H2MOD/GUI/imgui_integration/MOTD.cpp:167]: this could be a helper function

**Pull request merged**